# Data Wrangling - WeRateDogs Twitter Project

DESCRIPTION/COMMENTS

## Gather

In [1]:
# Import all packages needed or expected for this project
import pandas as pd
import numpy as np
import tweepy
import requests
import matplotlib.pyplot as plt
import tweepy
from tweepy import OAuthHandler
import json
from timeit import default_timer as timer
% matplotlib inline

In [2]:
# Begin by reading the provided file (file in hand)
twitter_arch = pd.read_csv('twitter_archive_enhanced.csv')

In [3]:
# View the first few lines to confirm and begin to get familiar
twitter_arch.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  8.924206e+17                    NaN                  NaN   
1  8.921774e+17                    NaN                  NaN   
2  8.918152e+17                    NaN                  NaN   
3  8.916896e+17                    NaN                  NaN   
4  8.913276e+17                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [4]:
twitter_arch['tweet_id'].head()

0    8.924206e+17
1    8.921774e+17
2    8.918152e+17
3    8.916896e+17
4    8.913276e+17
Name: tweet_id, dtype: float64

In [5]:
twitter_arch['tweet_id'] = twitter_arch['tweet_id'].astype(np.int64)

In [6]:
twitter_arch['tweet_id'].head()

0    892420643555336064
1    892177421306342912
2    891815181378083840
3    891689557279857920
4    891327558926688128
Name: tweet_id, dtype: int64

In [7]:
twitter_arch['tweet_id'] = twitter_arch['tweet_id'].astype(str)

In [8]:
twitter_arch['tweet_id'].head()

0    892420643555336064
1    892177421306342912
2    891815181378083840
3    891689557279857920
4    891327558926688128
Name: tweet_id, dtype: object

In [9]:
twitter_arch['tweet_id'][0]

'892420643555336064'

In [10]:
# Get and read in the file provided for Udacity's image predictions
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
file_name = 'image_predictions.tsv'
r = requests.get(url)

with open(file_name, 'wb') as fd:
    for chunk in r.iter_content(chunk_size=1024):
        fd.write(chunk)

In [12]:
image_pred = pd.read_csv('image_predictions.tsv', sep='\t')

In [13]:
image_pred.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [14]:
image_pred.shape

(2075, 12)

In [ ]:
# Final step for gathering - use the following steps to scrape Twitter
# for additional tweet data.

# Query Twitter API for each tweet in the Twitter archive and save JSON in a text file
# These are hidden to comply with Twitter's API terms and conditions
consumer_key = 'HIDDEN'
consumer_secret = 'HIDDEN'
access_token = 'HIDDEN'
access_secret = 'HIDDEN'

auth = OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True)

# NOTE TO STUDENT WITH MOBILE VERIFICATION ISSUES:
# df_1 is a DataFrame with the twitter_archive_enhanced.csv file. You may have to
# change line 17 to match the name of your DataFrame with twitter_archive_enhanced.csv
# NOTE TO REVIEWER: this student had mobile verification issues so the following
# Twitter API code was sent to this student from a Udacity instructor
# Tweet IDs for which to gather additional data via Twitter's API
tweet_ids = twitter_arch.tweet_id.values
len(tweet_ids)

# Query Twitter's API for JSON data for each tweet ID in the Twitter archive
count = 0
fails_dict = {}
start = timer()
# Save each tweet's returned JSON as a new line in a .txt file
with open('tweet_json.txt', 'w') as outfile:
    # This loop will likely take 20-30 minutes to run because of Twitter's rate limit
    for tweet_id in tweet_ids:
        count += 1
        print(str(count) + ": " + str(tweet_id))
        try:
            tweet = api.get_status(tweet_id, tweet_mode='extended')
            print("Success")
            json.dump(tweet._json, outfile)
            outfile.write('\n')
        except tweepy.TweepError as e:
            print("Fail")
            fails_dict[tweet_id] = e
            pass
end = timer()
print(end - start)
print(fails_dict)

In [15]:
# Use the collected txt file to obtain additional twitter information.

with open('tweet-json.txt') as json_file:
    # Create the dataframe to hold the data, with columns we will keep.
    tweet_json = pd.DataFrame(columns = ['tweet_id', 
                                        'favorites', 
                                      'retweets'])
    #Read in the lines and append for the variables we are storing
    for line in json_file:
        tweet = json.loads(line)
        tweet_json = tweet_json.append({
            'tweet_id': tweet['id'],
            'favorites': tweet['favorite_count'],
            'retweets': tweet['retweet_count']
        }, ignore_index=True)

In [16]:
tweet_json.head()

tweet_id favorites retweets
0  892420643555336193     39467     8853
1  892177421306343426     33819     6514
2  891815181378084864     25461     4328
3  891689557279858688     42908     8964
4  891327558926688256     41048     9774

In [17]:
all_columns = pd.Series(list(twitter_arch) + list(tweet_json) + list(image_pred))
all_columns[all_columns.duplicated()]

17    tweet_id
20    tweet_id
dtype: object

In [16]:
tweet_json.shape

(2354, 3)

## Assess

### Visual Assessment

The three dataframes that will be wrangled during this project are: twitter_arch, image_pred, and tweet_json.

In [17]:
twitter_arch.head(10)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336064                    NaN                  NaN   
1  892177421306342912                    NaN                  NaN   
2  891815181378083840                    NaN                  NaN   
3  891689557279857920                    NaN                  NaN   
4  891327558926688128                    NaN                  NaN   
5  891087950875896832                    NaN                  NaN   
6  890971913173991040                    NaN                  NaN   
7  890729181411236864                    NaN                  NaN   
8  890609185150311936                    NaN                  NaN   
9  890240255349198080                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   
5  2017-07-29 00:08:17 +0000   
6  2017-07-28 16:27:12 +0000   
7  2017-07-28 00:22:40 +0000   
8  2017-07-27 16:25:51 +0000   
9  2017-07-26 15:59:51 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   
5  <a href="http://twitter.com/download/iphone" r...   
6  <a href="http://twitter.com/download/iphone" r...   
7  <a href="http://twitter.com/download/iphone" r...   
8  <a href="http://twitter.com/download/iphone" r...   
9  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   
5  Here we have a majestic great white breaching ...                  NaN   
6  Meet Jax. He enjoys ice cream so much he gets ...                  NaN   
7  When you watch your owner call another dog a g...                  NaN   
8  This is Zoey. She doesn't want to be one of th...                  NaN   
9  This is Cassie. She is a college pup. Studying...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   
5                       NaN                        NaN   
6                       NaN                        NaN   
7                       NaN                        NaN   
8                       NaN                        NaN   
9                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   
5  https://twitter.com/dog_rates/status/891087950...                13   
6  https://gofundme.com/ydvmve-surgery-for-jax,ht...                13   
7  https://twitter.com/dog_rates/status/890729181...                13   
8  https://twitter.com/dog_rates/status/890609185...                13   
9  https://twitter.com/dog_rates/status/890240255...    

In [33]:
twitter_arch.tail()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2351  666049248165821952                    NaN                  NaN   
2352  666044226329799936                    NaN                  NaN   
2353  666033412701031936                    NaN                  NaN   
2354  666029285002620032                    NaN                  NaN   
2355  666020888022790016                    NaN                  NaN   

                      timestamp  \
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
2351  https://twitter.com/dog_rates/status/666049248...                 5   
2352  https://twitter.com/dog_rates/status/666044226...                 6   
2353  https://twitter.com/dog_rates/status/666033412...                 9   
2354  https://twitter.com/dog_rates/status/666029285...                 7   
2355  https://twitter.com/dog_rates/status/666020888...                 8   

      rating_denominator  name doggo floofer pupper puppo  
2351                  10  None  None    None   None  None  
2352                  10     a  None    None   None  None  
2353                  10     a  None    None   None  None  
2354                  10     a  None    None   None  None  
2355                  10  None  None    None   None  None

In [110]:
image_pred.head(10)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5  666050758794694657  https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6  666051853826850816  https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7  666055525042405380  https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8  666057090499244032  https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9  666058600524156928  https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   
5        1    Bernese_mountain_dog  0.651137    True    English_springer   
6        1              box_turtle  0.933012   False          mud_turtle   
7        1                    chow  0.692517    True     Tibetan_mastiff   
8        1           shopping_cart  0.962465   False     shopping_basket   
9        1        miniature_poodle  0.201493    True            komondor   

    p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0  0.156665    True            Shetland_sheepdog  0.061428    True  
1  0.074192    True          Rhodesian_ridgeback  0.072010    True  
2  0.138584    True                   bloodhound  0.116197    True  
3  0.360687    True           miniature_pinscher  0.222752    True  
4  0.243682    True                     Doberman  0.154629    True  
5  0.263788    True   Greater_Swiss_Mountain_dog  0.016199    True  
6  0.045885   False                     terrapin  0.017885   False  
7  0.058279    True                     fur_coat  0.054449   False  
8  0.014594   False             golden_retriever  0.007959    True  
9  0.192305    True  soft-coated_wheaten_terrier  0.082086    True

In [111]:
image_pred.tail()

tweet_id                                          jpg_url  \
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num           p1   p1_conf  p1_dog                  p2   p2_conf  \
2070        2       basset  0.555712    True    English_springer  0.225770   
2071        1  paper_towel  0.170278   False  Labrador_retriever  0.168086   
2072        1    Chihuahua  0.716012    True            malamute  0.078253   
2073        1    Chihuahua  0.323581    True            Pekinese  0.090647   
2074        1       orange  0.097049   False               bagel  0.085851   

      p2_dog                           p3   p3_conf  p3_dog  
2070    True  German_short-haired_pointer  0.175219    True  
2071    True                      spatula  0.040836   False  
2072    True                       kelpie  0.031379    True  
2073    True                     papillon  0.068957    True  
2074   False                       banana  0.076110   False

In [48]:
tweet_json.head()

tweet_id favorites retweets
0  892420643555336193     39467     8853
1  892177421306343426     33819     6514
2  891815181378084864     25461     4328
3  891689557279858688     42908     8964
4  891327558926688256     41048     9774

In [49]:
tweet_json.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
tweet_id     2354 non-null object
favorites    2354 non-null object
retweets     2354 non-null object
dtypes: object(3)
memory usage: 55.2+ KB


##### Initial Visual Assessments
1. twitter_arch has at least four columns that should be one (single variable for doggo, floofer, pupper, puppo)
2. twitter_arch dog type variable has many entries as 'None' rather than NaN
3. twitter_arch name field also has 'None' values; missing dog names
4. twitter_arch rating_numerator does not always follow the >10 method
5. image_pred p1 column does not have consistent formatting
6. image_pred p1 and others; not all predictions are dogs
7. image_pred column names are not very clear
8. twitter_arch in_reply_to columns are mostly responses to previous posts or additional comments




### Programmatic Assessment

In [18]:
twitter_arch.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null object
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(2), o

In [19]:
twitter_arch.nunique()

tweet_id                      2356
in_reply_to_status_id           55
in_reply_to_user_id             31
timestamp                     2356
source                           4
text                          2356
retweeted_status_id            121
retweeted_status_user_id        25
retweeted_status_timestamp     181
expanded_urls                 2218
rating_numerator                40
rating_denominator              18
name                           957
doggo                            2
floofer                          2
pupper                           2
puppo                            2
dtype: int64

In [20]:
twitter_arch['doggo'].value_counts()

None     2259
doggo      97
Name: doggo, dtype: int64

In [21]:
twitter_arch['pupper'].value_counts()

None      2099
pupper     257
Name: pupper, dtype: int64

In [22]:
twitter_arch.describe()

in_reply_to_status_id  in_reply_to_user_id  retweeted_status_id  \
count           7.800000e+01         7.800000e+01         1.810000e+02   
mean            7.455128e+17         2.015385e+16         7.720221e+17   
std             7.583419e+16         1.253546e+17         6.236131e+16   
min             6.660000e+17         1.185634e+07         6.660000e+17   
25%             6.760000e+17         3.086374e+08         7.190000e+17   
50%             7.035000e+17         4.196984e+09         7.800000e+17   
75%             8.260000e+17         4.196984e+09         8.200000e+17   
max             8.860000e+17         8.410000e+17         8.870000e+17   

       retweeted_status_user_id  rating_numerator  rating_denominator  
count              1.810000e+02       2356.000000         2356.000000  
mean               1.241437e+16         13.126486           10.455433  
std                9.597227e+16         45.876648            6.745237  
min                7.832140e+05          0.000000            0.000000  
25%                4.196984e+09         10.000000           10.000000  
50%                4.196984e+09         11.000000           10.000000  
75%                4.196984e+09         12.000000           10.000000  
max                7.870000e+17       1776.000000          170.000000

In [118]:
image_pred['p1'].value_counts()

golden_retriever             150
Labrador_retriever           100
Pembroke                      89
Chihuahua                     83
pug                           57
chow                          44
Samoyed                       43
toy_poodle                    39
Pomeranian                    38
cocker_spaniel                30
malamute                      30
French_bulldog                26
miniature_pinscher            23
Chesapeake_Bay_retriever      23
seat_belt                     22
Staffordshire_bullterrier     20
German_shepherd               20
Siberian_husky                20
Cardigan                      19
web_site                      19
Eskimo_dog                    18
Maltese_dog                   18
teddy                         18
Shetland_sheepdog             18
beagle                        18
Shih-Tzu                      17
Lakeland_terrier              17
Rottweiler                    17
kuvasz                        16
Italian_greyhound             16
          

In [119]:
tweet_json.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
tweet_id     2354 non-null object
favorites    2354 non-null object
retweets     2354 non-null object
dtypes: object(3)
memory usage: 55.2+ KB


In [120]:
image_pred.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


##### Initial Programmatic Assessments
1. twitter_arch has several retweeted or reply tweets included
2. twitter_arch dog type variables have very little data (majority is 'None')
3. twitter_arch contains some incorrect types (id columns as floats
4. twitter_arch rating_denominator is not consistent (not 10 throughout) and has at least one outlier (170); rating_numerator has some extreme outliers (1776 - a July 4th puppy?)
5. tweet_json data types are incorrect for favorites/retweets
6. each dataframe has a common key (tweet_id) and should be combined as one dataframe with only relevant variables

#### Assessment Summary

##### Quality Issues
1.	Dataframes do not share the same number of observations; missing or duplicated data
2.	twitter_arch has several retweeted or reply tweets included (should be excluded)
3. twitter_arch in_reply_to columns are mostly responses to previous posts or additional comments (should likely be excluded)
3.  twitter_arch tweet_id has duplicated ids
4.	Data types are incorrect within the twitter_arch and tweet_json dataframes
5.	twitter_arch contains some incorrect types (i.e. id columns as floats)
6.	tweet_json data types are incorrect for favorites/retweets
7.	twitter_arch dog type variables have very little data (majority are 'None', rather than NaN)
8.	twitter_arch name field also has 'None' values; missing or incorrect dog names
9.	image_pred p1, p2, p3 columns do not have consistent formatting (i.e. capitalization)
10.	twitter_arch rating_denominator is not consistent (not 10 throughout) and has at least one outlier (170); rating_numerator has some extreme outliers (1776 - a July 4th puppy?)
11.	Multiple columns/variables do not add needed value 
12.	image_pred column names are not very clear


##### Tidiness Issues
1.	Each dataframe has a common key (tweet_id) and should be combined as one dataframe
2.	twitter_arch has at least four columns that should be one (single variable for doggo, floofer, pupper, puppo)


## Clean
Before beginning any cleaning steps, create copies to ensure no data is lost.

In [40]:
twitter_arch_clean = twitter_arch.copy()
image_pred_clean = image_pred.copy()
tweet_json_clean = tweet_json.copy()

#### Define
1. Address the issue of duplicate tweet_ids in the twitter_arch file. Remove the retweets and responses as the content is not likely to be related to our analysis and may add in unnecessary entries or duplicates.
2. Change the 'None' values to NaN in the dog-type columns; if enough value to be useful, melt the variables into a single column.
3. Merge the three data frames, using the 'tweet_id' as our key.
4. 


In [41]:
all_columns = pd.Series(list(twitter_arch_clean) + list(tweet_json_clean) + list(image_pred_clean))
all_columns[all_columns.duplicated()]

17    tweet_id
20    tweet_id
dtype: object

#### Code

In [42]:
# Remove the retweets first 
twitter_arch_clean = twitter_arch_clean[twitter_arch_clean['retweeted_status_id'].isnull()]

In [43]:
twitter_arch_clean.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336064                    NaN                  NaN   
1  892177421306342912                    NaN                  NaN   
2  891815181378083840                    NaN                  NaN   
3  891689557279857920                    NaN                  NaN   
4  891327558926688128                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [44]:
# Next, remove the replies and response tweets
twitter_arch_clean = twitter_arch_clean[twitter_arch_clean['in_reply_to_status_id'].isnull()]

In [45]:
# Drop the columns that are no longer needed as only null values exist
# for these variables now
retweet_columns = ['retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp', 'in_reply_to_status_id', 'in_reply_to_user_id']
twitter_arch_clean = twitter_arch_clean.drop(retweet_columns, axis=1)

In [46]:
# Confirm that columns and rows have been removed
twitter_arch_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2097 non-null object
timestamp             2097 non-null object
source                2097 non-null object
text                  2097 non-null object
expanded_urls         2094 non-null object
rating_numerator      2097 non-null int64
rating_denominator    2097 non-null int64
name                  2097 non-null object
doggo                 2097 non-null object
floofer               2097 non-null object
pupper                2097 non-null object
puppo                 2097 non-null object
dtypes: int64(2), object(10)
memory usage: 213.0+ KB


In [47]:
twitter_arch_clean['doggo'] = twitter_arch_clean['doggo'].replace('None', np.NaN)
twitter_arch_clean['floofer'] = twitter_arch_clean['floofer'].replace('None', np.NaN)
twitter_arch_clean['pupper'] = twitter_arch_clean['pupper'].replace('None', np.NaN)
twitter_arch_clean['puppo'] = twitter_arch_clean['puppo'].replace('None', np.NaN)

In [48]:
twitter_arch_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2097 non-null object
timestamp             2097 non-null object
source                2097 non-null object
text                  2097 non-null object
expanded_urls         2094 non-null object
rating_numerator      2097 non-null int64
rating_denominator    2097 non-null int64
name                  2097 non-null object
doggo                 83 non-null object
floofer               10 non-null object
pupper                230 non-null object
puppo                 24 non-null object
dtypes: int64(2), object(10)
memory usage: 213.0+ KB


Considering the low # of actual entries for the dog type variable, it seems that the best option is to remove these columns as they do not add enough value to warrant including in our future analysis.

In [49]:
# Drop the four dog-type columns
type_columns = ['doggo', 'pupper', 'floofer', 'puppo']
twitter_arch_clean = twitter_arch_clean.drop(type_columns, axis=1)

In [50]:
twitter_arch_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 8 columns):
tweet_id              2097 non-null object
timestamp             2097 non-null object
source                2097 non-null object
text                  2097 non-null object
expanded_urls         2094 non-null object
rating_numerator      2097 non-null int64
rating_denominator    2097 non-null int64
name                  2097 non-null object
dtypes: int64(2), object(6)
memory usage: 147.4+ KB


In [51]:
twitter_arch_clean.head(2)

tweet_id                  timestamp  \
0  892420643555336064  2017-08-01 16:23:56 +0000   
1  892177421306342912  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   

   rating_denominator     name  
0                  10  Phineas  
1                  10    Tilly

In [52]:
tweet_json_clean['tweet_id'].head(2)

0    892420643555336193
1    892177421306343426
Name: tweet_id, dtype: object

In [53]:
twitter_arch_clean['tweet_id'].head(2)

0    892420643555336064
1    892177421306342912
Name: tweet_id, dtype: object

In [55]:
image_pred_clean['tweet_id'].head(2)

0    666020888022790149
1    666029285002620928
Name: tweet_id, dtype: int64

In [56]:
image_pred_clean['tweet_id'] = image_pred_clean['tweet_id'].astype(str)

In [57]:
image_pred_clean['tweet_id'].head(2)

0    666020888022790149
1    666029285002620928
Name: tweet_id, dtype: object

# Error Issues with Merge

In [58]:
# Merge the three dataframes as they all relate to a single tweet_id
# Begin by merging twitter_arch_clean with tweet_json_clean

twitter_arch_clean = pd.merge(twitter_arch_clean, tweet_json_clean, how = 'left', on = ['tweet_id'])

In [59]:
twitter_arch_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2096
Data columns (total 10 columns):
tweet_id              2097 non-null object
timestamp             2097 non-null object
source                2097 non-null object
text                  2097 non-null object
expanded_urls         2094 non-null object
rating_numerator      2097 non-null int64
rating_denominator    2097 non-null int64
name                  2097 non-null object
favorites             0 non-null object
retweets              0 non-null object
dtypes: int64(2), object(8)
memory usage: 180.2+ KB


#### Test